std/experimental/allocator/building_blocks/allocator_list.d

```diff
@@ -0,0 +1,614 @@
+        {
+            if (n.owns(b) == Ternary.yes) return n.expand(b, delta);
+        }
+        return false;
+    }
+
+    /**
+    Defined only if $(D Allocator.reallocate) is defined. Finds the owner of
+    $(D b) and calls $(D reallocate) for it. If that fails, calls the global
+    $(D reallocate), which allocates a new block and moves memory.
+    */
+    static if (hasMember!(Allocator, "reallocate"))
+    bool reallocate(ref void[] b, size_t s)
+    {
+        // First attempt to reallocate within the existing node
+        if (!b.ptr)
+        {
+            b = allocate(s);
+            return b.length == s;
+        }
+        for (auto p = &root, n = *p; n; p = &n.next, n = *p)
+        {
+            if (n.owns(b) == Ternary.yes) return n.reallocate(b, s);
+        }
+        // Failed, but we may find new memory in a new node.
+        return .reallocate(this, b, s);
+    }
+
+    /**
+     Defined if $(D Allocator.deallocate) and $(D Allocator.owns) are defined.
+    */
+    static if (hasMember!(Allocator, "deallocate")
+        && hasMember!(Allocator, "owns"))
+    bool deallocate(void[] b)
+    {
+        if (!b.ptr) return true;
+        assert(allocators.length);
+        assert(owns(b) == Ternary.yes);
+        bool result;
+        for (auto p = &root, n = *p; ; p = &n.next, n = *p)
+        {
+            assert(n);
+            if (n.owns(b) != Ternary.yes) continue;
+            result = n.deallocate(b);
+            // Bring to front
+            if (n != root)
+            {
+                *p = n.next;
+                n.next = root;
+                root = n;
+            }
+            if (n.empty != Ternary.yes) return result;
+            break;
+        }
+        // Hmmm... should we return this allocator back to the wild? Let's
+        // decide if there are TWO empty allocators we can release ONE. This
+        // is to avoid thrashing.
+        // Note that loop starts from the second element.
+        for (auto p = &root.next, n = *p; n; p = &n.next, n = *p)
+        {
+            if (n.unused || n.empty != Ternary.yes) continue;
+            // Used and empty baby, nuke it!
+            n.a.destroy;
+            *p = n.next;
+            n.setUnused;
+            break;
+        }
+        return result;
+    }
+
+    /**
+    Defined only if $(D Allocator.owns) and $(D Allocator.deallocateAll) are
+    defined.
+    */
+    static if (ouroboros && hasMember!(Allocator, "deallocateAll")
+        && hasMember!(Allocator, "owns"))
+    bool deallocateAll()
+    {
+        Node* special;
+        foreach (ref n; allocators)
+        {
+            if (n.unused) continue;
+            if (n.owns(allocators) == Ternary.yes)
+            {
+                special = &n;
+                continue;
+            }
+            n.a.deallocateAll;
+            n.a.destroy;
+        }
+        assert(special || !allocators.ptr);
+        if (special)
+        {
+            special.deallocate(allocators);
```

Comment:
I think this should be `deallocateAll`, otherwise any memory left in the special node's allocator that was not used up as storage for `allocators` will never be freed.